File: app/src/main/java/com/mirash/passkeeper/activity/pin/view/PinCircleView.java
package com.mirash.passkeeper.activity.pin.view;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.widget.Checkable;
import android.widget.FrameLayout;
import android.widget.ImageView;

import androidx.annotation.NonNull;

import com.mirash.passkeeper.R;

/**
 * @author Mirash
 */
public class PinCircleView extends FrameLayout implements Checkable {
    private boolean isChecked;
    private ImageView checkedImageView;
    private ImageView idleImageView;
    private ObjectAnimator animator;

    public PinCircleView(@NonNull Context context) {
        super(context);
        init();
    }

    private void init() {
        idleImageView = new ImageView(getContext());
        idleImageView.setImageResource(R.drawable.ic_circle_outline);
        checkedImageView = new ImageView(getContext());
        checkedImageView.setImageResource(R.drawable.ic_circle_filled);
        checkedImageView.setImageAlpha(0);
        addView(idleImageView);
        addView(checkedImageView);
        animator = new ObjectAnimator().setDuration(250);
        animator.addUpdateListener(valueAnimator -> {
            int value = (int) valueAnimator.getAnimatedValue();
            checkedImageView.setImageAlpha(value);
        });
    }


    @Override
    public void setChecked(boolean checked) {
        if (isChecked != checked) {
            isChecked = checked;
            if (checked) {
                animator.setIntValues(0, 255);
            } else {
                animator.setIntValues(255, 0);
            }
            animator.start();

        }
    }

    @Override
    public boolean isChecked() {
        return isChecked;
    }

    @Override
    public void toggle() {
        setChecked(!isChecked);
    }
}

File: app/src/main/java/com/mirash/familiar/activity/pin/view/PinCircleView.kt
package com.mirash.familiar.activity.pin.view

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.content.Context
import android.widget.Checkable
import android.widget.FrameLayout
import android.widget.ImageView
import com.mirash.familiar.R

/**
 * @author Mirash
 */
class PinCircleView(context: Context) : FrameLayout(context), Checkable {
    private var isChecked = false
    private var checkedImageView: ImageView
    private var idleImageView: ImageView
    private var animator: ObjectAnimator

    init {
        idleImageView = ImageView(context)
        idleImageView.setImageResource(R.drawable.ic_circle_outline)
        checkedImageView = ImageView(context)
        checkedImageView.setImageResource(R.drawable.ic_circle_filled)
        checkedImageView.imageAlpha = 0
        addView(idleImageView)
        addView(checkedImageView)
        animator = ObjectAnimator().setDuration(250)

        ObjectAnimator().setDuration(250).let {
            it.addUpdateListener { valueAnimator: ValueAnimator ->
                checkedImageView.imageAlpha = valueAnimator.getAnimatedValue() as Int
            }
            animator = it
        }
    }

    override fun setChecked(checked: Boolean) {
        if (isChecked != checked) {
            isChecked = checked
            if (checked) {
                animator.setIntValues(0, 255)
            } else {
                animator.setIntValues(255, 0)
            }
            animator.start()
        }
    }

    override fun isChecked(): Boolean {
        return isChecked
    }

    override fun toggle() {
        setChecked(!isChecked)
    }
}
